File: java/leetcode/SpecialBinaryString.java
//Leetcode problem 761 Speical Binary String
//Solution written by Xuqiang Fang on 12 July, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
class Solution{
    public String makeLargestSpecial(String S) {
        List<String> ans = new ArrayList<>();
        int prev = 0, count = 0;
        for(int i=0; i<S.length(); ++i){
            if(S.charAt(i) == '1'){
                count++;
            } 
            else{
                count--;
            }
            if(count == 0){
                ans.add('1' + makeLargestSpecial(S.substring(prev+1, i))+'0');
                prev = i+1;
            }
        }
        ans.sort(new Comp());
        StringBuilder sb = new StringBuilder();
        for(String a : ans){
            sb.append(a);
        }
        return sb.toString();
    }
    class Comp implements Comparator<String>{
        public int compare(String a, String b){
            int l = Math.min(a.length(), b.length());
            for(int i=0; i<l; ++i){
                if(a.charAt(i) != b.charAt(i)){
                    return (int)(b.charAt(i)-a.charAt(i));
                }
            }
            return b.length() - a.length();
        }
    }
}

public class SpecialBinaryString{
	public static void main(String[] args){
		Solution s = new Solution();
        String S = "11011000";
        System.out.println(s.makeLargestSpecial(S));
	}
}

File: py/leetcode/SpecialBinaryString.py
'''
Leetcode problem No 761 Speical Binary String
Solution written by Xuqiang Fang on 11 July, 2018
'''
import functools
def mycomp(a, b):
    l = min(len(a), len(b))
    for i in range(l):
        if a[i] != b[i]:
            if int(a[i])-int(b[i]) > 0:
                return -1
            return 1
    return len(a)-len(b)

class Solution(object):
    def makeLargestSpecial(self, S):
        """
        :type S: str
        :rtype: str
        """
        ans = []
        prev = 0; c = 0
        for i, v in enumerate(S):
            c += 1 if v == '1' else -1
            if c == 0:
                ans.append('1{}0'.format(self.makeLargestSpecial(S[prev+1: i])))
                prev = i + 1
        ans.sort(key = functools.cmp_to_key(mycomp))
        return ''.join(ans)
def main():
    s = Solution()
    S = "11011000"
    print(s.makeLargestSpecial(S))
    S = '101011110000101010111111000000'
    print(s.makeLargestSpecial(S))
    S = '111010011000'
    print(s.makeLargestSpecial(S))
    S = "101011001110011010001111100000"
    print(s.makeLargestSpecial(S))
    print("111110000011101001100011001010")
main()
